drivers/flash/flash_stm32l4x.c

```diff
@@ -244,6 +245,159 @@ int flash_stm32_write_range(const struct device *dev, unsigned int offset,
 	return rc;
 }
 
+static __unused int write_optb(const struct device *dev, uint32_t mask,
+			       uint32_t value)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	int rc;
+
+	if (regs->CR & FLASH_CR_OPTLOCK) {
+		return -EIO;
+	}
+
+	if ((regs->OPTR & mask) == value) {
+		return 0;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->OPTR = (regs->OPTR & ~mask) | value;
+	regs->CR |= FLASH_CR_OPTSTRT;
+
+	/* Make sure previous write is completed. */
+	barrier_dsync_fence_full();
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_FLASH_STM32_WRITE_PROTECT)
+
+/*
+ * STM32L4 allows for 2 write protected memory areas, c.f. FLASH_WEP1AR, FLASH_WRP1BR
+ * which are defined by their start and end pages, i.e. there are two consecutive ranges
+ * the current API passes in two 32 bit mask struct flash_stm32_ex_op_sector_wp_in,
+ * which are barely sufficient to encode the 256 2k pages for the L451 (512kb Flash) part
+ * (embedd all 4 start / end pages number for the two protected areas, each using 8 bits)
+ * this would then be different to STM32F4 meaning of those bits, where they encode single
+ * pages. This could be done here as well, the consecutivity check must be implemented and
+ * thins may be a bit complex regarding the assignement of ranges to FLASH_WEP1AR,
+ * FLASH_WRP1BR
```

Comment:
Instead of implementing complex and error prone translation layer, you should add separate write protect extended operation for stm32 chips which have range based write protect.

Reply:
something like FLASH_STM32_EX_OP_SECTOR_WP_RANGED in include/zephyr/drivers/flash/stm32_flash_api_extensions.h?

Reply:
Yes.